### PR TITLE
Fixed floating point issue with asinh function

### LINKED ISSUE #63271: f32/64::asinh() should return -0.0 for -0.0
Currently, both of the following return false:

```rust
(-0.0 as f64).asinh().is_sign_negative()
(-0.0 as f32).asinh().is_sign_negative()
```

However, in other languages like [C](https://en.cppreference.com/w/c/numeric/math/asinh), `asinh(-0.0)` would always return `-0.0`. We might need another fast case for 0.0 and -0.0 here: https://github.com/rust-lang/rust/blob/f01b9f803b59f170f5dabaaa8aedc96abe45bfea/src/libstd/f64.rs#L833-L839

### PR DESCRIPTION
**Pull request by Phosphorus15**
This should fixes #63271 , in which `asinh(-0.0)` returns `0.0` instead of `-0.0`.
according to @nagisa 
> 
> 
> IEEE-754 (2008), section 9.2.1:
> 
> > For the functions expm1, exp2m1, exp10m1, logp1, log2p1, log10p1, sin, tan, sinPi, atanPi, asin, atan, sinh, tanh, asinh, and atanh, f(+0) is +0 and f(−0) is −0 with no exception.
> 
> and
> 
> > sinh(±∞) and asinh(±∞) are ±∞ with no exception.

After ensuring that the function `asinh` is the only function affected (functions like `sin`, `sinh` are all based on `cmath` library or `llvm` intrinsics), and that `atanh` always gives the correct result. The only function to modify is `asinh`.

### CODE REVIEW

**Comment by Mark-Simulacrum** [comment]
r? @nagisa perhaps?

Not sure who a good review here would be.

**Comment by nagisa** [comment]
You’ll probably want to use [copysign](https://doc.rust-lang.org/stable/std/primitive.f32.html#method.copysign) instead and not do computations in `abs(x)` space only.

File: src/libstd/f32.rs [hunk 1]
```diff
@@ -908,10 +908,9 @@ impl f32 {
     #[stable(feature = "rust1", since = "1.0.0")]
     #[inline]
     pub fn asinh(self) -> f32 {
-        if self == NEG_INFINITY {
-            NEG_INFINITY
-        } else {
-            (self + ((self * self) + 1.0).sqrt()).ln()
+        match self {
+            x if x == NEG_INFINITY => NEG_INFINITY,
+            x => (x + ((x * x) + 1.0).sqrt()).ln().copysign(self)
```
Comment (lzutao) [inline src/libstd/f32.rs]: I would prefer to use if-else directly.

Comment (Phosphorus15) [reply src/libstd/f32.rs]: Yea, that sounds reasonable, maybe consider changing the original `acosh` BTW?
```rust
pub fn acosh(self) -> f64 {
        match self {
            x if x < 1.0 => NAN,
            x => (x + ((x * x) - 1.0).sqrt()).ln(),
        }
    }
```

File: src/libstd/f64.rs [hunk 2]
```diff
@@ -481,16 +481,16 @@ impl f64 {
     #[stable(feature = "rust1", since = "1.0.0")]
     #[inline]
     #[rustc_deprecated(since = "1.10.0",
-                       reason = "you probably meant `(self - other).abs()`: \
+    reason = "you probably meant `(self - other).abs()`: \
```
Comment (nagisa) [inline src/libstd/f64.rs]: Why?

**Review by nagisa — APPROVED** [review]

File: src/libstd/f64.rs (hunk 2 above)
Comment (Phosphorus15) [reply src/libstd/f64.rs]: Sorry for changing the identation 😢 , it might because of my IDE's formatter.

**Pull request merged**